### DISCUSSION TITLE
termination step is skipped due to large time step in Events

### DISCUSSION THREAD

**Discussion by leigedove** [pybamm-team/PyBaMM#3895] [Q&A]
In handling events and termination conditions in PyBaMM, there can be issues with correctly triggering the termination conditions due to too large a time step. (i.e. for instance, the termination voltage condition should be met at a certain moment, but due to the choice of time step, this condition is “skipped”). Are there any good solutions to address this problem?

```
import pybamm
import numpy as np

model = pybamm.lithium_ion.DFN()
parameter_values = pybamm.ParameterValues("Chen2020")

def voltage_cutoff(variables):
    return 2.9 - variables["Terminal voltage [V]"]

voltage_termination = pybamm.step.CustomTermination(
    name="Voltage cutoff", event_function=voltage_cutoff)

terminations = [voltage_termination]

t = np.linspace(0, 50, 600) # this time step is too large and skip the termination condition
sin_t = -15 * np.sin(0.5*np.pi * t)
custom_current = np.column_stack([t, sin_t])
experiment = pybamm.Experiment([(pybamm.step.current(custom_current,termination=terminations))])

sim = pybamm.Simulation(model, parameter_values=parameter_values, experiment=experiment)
sim.solve(initial_soc=0)
sim.plot()
```

I tried adding some extra logic to the custom voltage cutoff event function to make the function gradually decrease the range of return values as it nears that value.

```
def voltage_cutoff(variables):
    cutoff_voltage = 2.9
    current_voltage = variables["Terminal voltage [V]"]
    buffer_range = 0.05  
    if current_voltage < cutoff_voltage and current_voltage >= cutoff_voltage - buffer_range:
        return (cutoff_voltage - current_voltage) / buffer_range
    else:
        return cutoff_voltage - current_voltage
```

But there is an error: “Boolean operator not defined for Symbols.”

Thanks!!!

**Comment by rtimms** [comment]
For the casadi solver you need to set the max step size to match the step size in the drive cycle e.g.
```
    # get smallest step size in UDDS
    dt = np.diff(udds[:, 0]).min()
    # define experiment
    experiment = pybamm.Experiment([pybamm.step.current(udds)])
    # create solver with max step size of dt
    solver = pybamm.CasadiSolver(extra_options_setup={"max_step_size": dt})
```

**Reply by leigedove** [reply]
Yes, thanks for your quick reply. But it still doesn't terminate when the voltage reaches 2.9 V
```
import pybamm
import numpy as np

model = pybamm.lithium_ion.DFN(options={"thermal": "lumped"})
parameter_values = pybamm.ParameterValues("Chen2020")

def voltage_cutoff(variables):
    return 2.9 - variables["Terminal voltage [V]"]

voltage_termination = pybamm.step.CustomTermination(
    name="Voltage cutoff", event_function=voltage_cutoff)

terminations = [voltage_termination]

t = np.linspace(0, 20, 600)
sin_t = -15 * np.sin(0.5 * np.pi * t)
custom_current = np.column_stack([t, sin_t])

experiment = pybamm.Experiment([(pybamm.step.current(custom_current, termination=terminations))])

dt = np.diff(custom_current[:, 0]).min()

solver = pybamm.CasadiSolver(extra_options_setup={"max_step_size": dt})

sim = pybamm.Simulation(model, parameter_values=parameter_values, experiment=experiment, solver=solver)
sim.solve(initial_soc=0)
sim.plot()

solution = sim.solution
print(solution["X-averaged cell temperature [K]"].data, solution.termination)
```

**Reply by leigedove** [reply]
Shall I make some changes in the event function? Thank you😊
def voltage_cutoff(variables):
    return 2.9 - variables["Terminal voltage [V]"]
    
 voltage_termination = pybamm.step.CustomTermination(
    name="Voltage cutoff", event_function=voltage_cutoff)

**Reply by rtimms** [reply]
can you define your event the other way around? i.e. `return V - 2.9` iirc the event needs to switch from positive to negative. 

also, this event is already built into the model by default, you just need to change the parameter “Lower voltage cut-off [V]”.

**Reply by rtimms** [reply]
See https://github.com/pybamm-team/PyBaMM/blob/8056d224e75b3926743319a4743abb7c6ab0dac2/pybamm/experiment/step/step_termination.py#L78

**Reply by leigedove** [reply]
Thank you for your reply. Yeah, right. but I'm doing a charging process and the initial soc = 0. The voltage is rising, so the event should be `return 2.9 - V`.

I am trying to use different termination conditions. This one is just an example. And if there are multiple steps in the experiment and each termination condition is different, I can't modify the cutoff voltage, right?

**Comment by martinjrobins** [comment]
You could try the [IDAKLU](https://docs.pybamm.org/en/latest/source/api/solvers/idaklu_solver.html) solver which has better event-handling capabilities (the solver, Sundials, handles all the events)

**Reply by leigedove** [reply]
When i use IDAKLU, there is alway an error: "except pybamm.SolverError as error: if (non-positive at initial conditions" in error.message and "[experiment]" in error.message): TypeError: 'NoneType' object is not iterable"

Is this because I used a custom current function?

**Reply by martinjrobins** [reply]
I tried the code above (https://github.com/pybamm-team/PyBaMM/discussions/3895#discussioncomment-8789267) with the IDAKLU solver and it worked without error. Is this the code you used?

**Comment by leigedove** [comment]
I find that when using a periodic current function like sin (with positive and negative currents), it misses the termination condition. But when using a current function that is only monotonic and only negative, it works.

**Reply by Thiyaga2025** [reply]
hi @leigedove , @rtimms 

i am running aging simulation for drive cycle conditions. The issue is the simulation doesn't stop at the termination voltage, instead it completes the full duration which leads to lower voltage error. the code is as below,
#
df_dc = pd.read_csv('WLTP_c3_LGM50_(av 1,62A dis, 0,53A cha),tab-delim.txt',sep="\t").to_numpy()
df_dc[:,1] *= -1
drive_step = pybamm.step.current(df_dc,  termination="3.5 V", duration="5 hours")
#
aging_experiment = pybamm.Experiment([
    (
    "Discharge at 1C for 3 minutes",
    drive_step,
    "Charge at 0.3C until 4.2 V",
    "Hold at 4.2 V until C/50",
    )
]
)
#
sim = pybamm.Simulation(model, parameter_values=parameter_values, experiment=aging_experiment,
                           solver=solver)
sol = sim.solve()
#
will be thankful for any help.
